### PR TITLE
International stemming

### PR DESCRIPTION
**Pull request by miceno**
It is a rather long PR.

Fortunately, we improved testing last weeks, and it helped refactoring, so now all tests pass and new tests were added.

Allow indexing content using current locale of the bundle:
* schema update.

Allow pluggable Stemmers:
* PECL stem
* English Snoball
* Wamania stemmer.

Search using all available cataloguing locales.
Update language detection code:
* convert Spanish and Catalan to utf-8
* add a command to test quality of language detection.

Update test database setup to allow unattended schema updates.

### CODE REVIEW

**Comment by miceno** [comment]
Hi! Did you have an opportunity to look at this? I am happy to receive feedback.

**Comment by collectiveaccess** [comment]
Sorry for the delay. I will give it a close look in the next day or two.

**Comment by miceno** [comment]
Hi!

Some weeks ago there were conflicts on this PR and I solved them.

A couple of days ago, there were no conflicts on this PR, but there are some now... I will solve them, but I suggest a review of this PR before accepting any other PR.

Regards.

**Comment by miceno** [comment]
More conflicts solved...

**Comment by miceno** [comment]
Any feedback on this?

**Comment by miceno** [comment]
Any feedback on this?

**Comment by miceno** [comment]
Giving up, I will fork and continue on my own fork.

**Comment by collectiveaccess** [comment]
Again, sorry that we've been distracted with things. Will review and get back to you by Monday.

**Comment by collectiveaccess** [comment]
On first examination it look good in general. I have installed it on my test machine and will try it out for a little bit. I also need to look more closely at the changes to SqlSearch.php, which are extensive and deserve additional attention.

**Comment by miceno** [comment]
I have resolved conflicts, and move the migration in the PR to number 165. Try it again.

I have to test it again, I will do by tomorrow.

**Comment by miceno** [comment]
I have also added some TODO to review undefined variables. Will you please also have a look at them?

**Comment by collectiveaccess** [comment]
Yes will do before Friday. I won't be around much today.

File: app/lib/Plugins/SearchEngine/SqlSearch.php [hunk 1]
```diff
@@ -91,12 +92,19 @@ public function __construct($po_db=null) {
...
+            $vs_stemmer = self::SNOBALL_STEMMER;
+        }
+        $vo_stemmer_factory = StemmerFactory::get_instance();
+		$this->opo_stemmer = $vo_stemmer_factory->create($vs_stemmer);
 		$this->opb_do_stemming = (int)trim($this->opo_search_config->get('search_sql_search_do_stemming')) ? true : false;
 		
 		$this->initDbStatements();
-		
+
+		// TODO: ops_indexing_tokenizer_regex is not used in this class
```
Comment (collectiveaccess) [inline app/lib/Plugins/SearchEngine/SqlSearch.php:107]: We can get rid of this. At one point we were tokenizing in some cases at this level, but that's not coming back.

File: app/lib/Plugins/SearchEngine/SqlSearch.php [hunk 2]
```diff
@@ -863,79 +871,86 @@ private function _doQueriesForSqlSearch($po_rewritten_query, $pn_subject_tablenu
...
+							        ";
+                                $pa_direct_sql_query_params = array(); // don't pass any params
+
+                                foreach ($va_temp_tables as $vs_temp_table) {
+                                    $this->_dropTempTable($vs_temp_table);
+                                }
+
+                                break;
+                            }
+                            // TODO: Missing break statement
```
Comment (collectiveaccess) [inline app/lib/Plugins/SearchEngine/SqlSearch.php:953]: It intentionally falls to the default case here when strictPhraseSearching is not in effect.

**Comment by collectiveaccess** [comment]
As you can see I've started a code review of the SqlSearch changes. In general it all makes sense and seems reasonable. We'll have to wire up updateIndexingInPlace() to support locales before we can start considering merging though.

**Comment by miceno** [comment]
Thank you for the review!

Concerning `updateIndexingInPlace`, the point is that if no locale is provided it will use the global one: see https://github.com/collectiveaccess/providence/pull/441/files#diff-cb738331901dc23c6f0bdf37ea30de3bR2099

I think it should use the locale of the attribute, but I am not sure how to get it on `updateIndexingInPlace`. Will you please advice?

**Comment by collectiveaccess** [comment]
Obviously indexing should honor the locale whenever present. Potentially, locale can be changes for an existing attribute, so we can't assume that the initial locale is still correct when updating the index. I think the only solution would be to change the indexer to pass locale_ids to updateIndexingInPlace.

**Comment by yelto** [comment]
Did you consider augmenting the index with language rather than with locale?

**Comment by collectiveaccess** [comment]
Closing for now.

**Pull request closed without merging**